plugin/pkg/admission/podpreset/exists/admission.go

```diff
@@ -0,0 +1,155 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exists
+
+import (
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/discovery"
+	kubeclientset "k8s.io/client-go/kubernetes"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog"
+	scadmission "github.com/kubernetes-incubator/service-catalog/pkg/apiserver/admission"
+)
+
+const (
+	// PluginName is name of admission plug-in
+	PluginName = "ServiceCatalogPodPresetExists"
+	// SupportedPodPresetVersion is the GroupVersion for what we require for
+	// creating PodPresetTemplates in bindings.
+	SupportedPodPresetVersion = "settings.k8s.io/v1alpha1"
+)
+
+func init() {
+	admission.RegisterPlugin(PluginName, func(io.Reader) (admission.Interface, error) {
+		return NewExists()
+	})
+}
+
+// exists is an implementation of admission.Interface.
+// It enforces that if a binding is being defined with a PodPresetTemplate
+// that the cluster actually has support for it.
+type exists struct {
+	*admission.Handler
+	discovery         discovery.DiscoveryInterface
+	sync.Mutex        // protects podPresetsChecked and podPresetsExists
+	podPresetsChecked bool
+	podPresetsExists  bool
+}
+
+var _ = scadmission.WantsKubeClientSet(&exists{})
+
+func (e *exists) supportsPodPresets() bool {
+	e.Lock()
+	defer e.Unlock()
+	return e.podPresetsExists
+}
+
+// checkClusterForPodPresets uses discovery to find if the cluster has the
+// right version of settings that supports PodPresets. It will return true
+// if we have successfully queried the Discovery API and hence we know
+// for sure if the cluster supports PodPresets or not.
+// In that case, it also sets podPresetsChecked to true and podPresetsExists
+// to true if the cluster supports PodPresets.
+func (e *exists) checkClusterForPodPresets() bool {
+	synced := false
+	e.Lock()
+	synced = e.podPresetsChecked
+	e.Unlock()
+	if synced {
+		return true
+	}
+
+	resourceList, err := e.discovery.ServerResourcesForGroupVersion(SupportedPodPresetVersion)
+
+	if err != nil && !errors.IsNotFound(err) {
+		// We don't know if the cluster supports PodPresets or not
+		glog.V(4).Infof("ServerResourcesForGroupVersion failed: %s", err)
+		return false
+
+	}
+
+	e.Lock()
+	defer e.Unlock()
+	e.podPresetsChecked = true
+
+	if err != nil && errors.IsNotFound(err) {
+		// No such resource, which means the cluster does not support PodPresets
+		glog.V(4).Infof("No PodPreset support in the cluster, not allowing PodPresetTemplates in Bindings")
+		return true
+	}
+
+	for _, resource := range resourceList.APIResources {
+		if resource.Name == "podpresets" {
+			glog.V(4).Infof("PodPreset support found, allowing PodPresetTemplates in Bindings")
+			e.podPresetsExists = true
+		}
+	}
+	return true
+}
+
+func (e *exists) Admit(a admission.Attributes) error {
```

Comment:
@derekwaynecarr @vaikas-google non nil error return is "okay to admit" ?

Comment here is non-specific.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/interfaces.go#L56